sklearn/pipeline.py

```diff
@@ -516,6 +531,109 @@ def _pairwise(self):
+        an estimator.
+
+    memory : Instance of joblib.Memory or string, optional (default=None)
+        Used to cache the output of the computation of the tree.
+        By default, the current directory is used to cache the estimators.
+        If a string is given, it is the path to the caching directory.
+
+    Attributes
+    ----------
+    named_steps : dict
+        Read-only attribute to access any step parameter by user given name.
+        Keys are step names and values are steps parameters.
+
+    Examples
+    --------
+    >>> from sklearn.datasets import samples_generator
+    >>> from sklearn.decomposition import PCA
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.externals.joblib import Memory
+    >>> from sklearn.model_selection import GridSearchCV
+    >>> from sklearn.pipeline import CachedPipeline
+    >>> from tempfile import mkdtemp
+    >>> # setup a temporary directory
+    >>> cachedir = mkdtemp()
+    >>> # setup the cache
+    >>> memory = Memory(cachedir=cachedir, verbose=0)
+    >>> # generate some data to play with
+    >>> X, y = samples_generator.make_classification(
+    ...     n_samples=100, n_informative=5, n_redundant=0, random_state=42)
+    >>> # Create a pipeline with a PCA and an RandomForestClassifier
+    >>> pca = PCA()
+    >>> clf = RandomForestClassifier()
+    >>> pipeline = CachedPipeline([('pca', pca), ('rf', clf)],
+    ...                           memory=memory)
+    >>> # Make a grid_search of the best parameters
+    >>> parameters = {'pca__n_components': (.25, .5),
+    ...               'rf__n_estimators': (10, 20)}
+    >>> # To avoid refitting the same PCA for different configuration
+    >>> # CachedPipeline will store the PCA estimator and load it
+    >>> # when required durin the grid search
+    >>> grid_search = GridSearchCV(pipeline, parameters, verbose=0)
+    >>> grid_search.fit(X, y)
+    ...                                              # doctest: +ELLIPSIS
+    GridSearchCV(...)
+
+    See also
+    --------
+    Pipeline
+        Pipeline of transforms with a final estimator.
+
+    """
+
+    def __init__(self, steps, memory=None):
+        self.memory = memory
+        super(CachedPipeline, self).__init__(steps)
+
+    def _fit_single_transform(self, transformer, name, idx_transform, X, y,
+                              **fit_params_trans):
+        memory = self.memory
+        if memory is None:
+            memory = Memory(cachedir=os.getcwd(), verbose=0)
+        elif isinstance(memory, six.string_types):
+            memory = Memory(cachedir=memory, verbose=0)
+        elif not isinstance(memory, Memory):
+            raise ValueError('memory is either `str` or a `joblib.Memory`'
+                             ' instance')
+
+        # Clone the transformer to maximize cache hits
+        clone_transformer = clone(transformer)
+        Xt, transform = memory.cache(_fit_transform_one)(
```

Comment:
Why "transform", I think we can shadow the previous value of the `transformer` variable. Alternatively you could use `fitted_transformer` to make it explicit.